### PR TITLE
Adds conditional create

### PR DESCRIPTION
**Pull request by brendankowitz**
## Description
- Adds support for conditional updates as described in the spec http://hl7.org/fhir/http.html#ccreate.
- Adds a Feature flag that can enable Conditional-Update conditionally in FHIR Server deployments
- Moves Feature Flags into their own Modules
- Sets the `conditionalCreate` capability in the capability statement

Limitations:
- Several race conditions have been identified where a record could be incorrectly inserted or updated after the conditional search has been performed

## Related issues
Addresses ab#63059

## Testing
- All existing tests should pass.
- Adds new Unit tests to validate Handler logic
- Adds new E2E tests with conditional update scenarios

### CODE REVIEW

File: src/Microsoft.Health.Fhir.Shared.Api/Modules/FeatureFlags/ConditionalCreates/ConditionalCreatePostConfigureOptions.cs [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+
+            _features = features.Value;
+            _configuredConformanceProvider = configuredConformanceProvider;
+        }
+
+        public void PostConfigure(string name, MvcOptions options)
+        {
+            if (_features.SupportsConditionalCreate)
+            {
+                // Turns on conditional updates, even when not configured in the DefaultCapabilities.json
```
Comment (namadabu) [inline src/Microsoft.Health.Fhir.Shared.Api/Modules/FeatureFlags/ConditionalCreates/ConditionalCreatePostConfigureOptions.cs]: nit: create

File: src/Microsoft.Health.Fhir.Shared.Core.UnitTests/Features/Resources/ResourceHandlerTests.cs [hunk 2]
```diff
@@ -63,13 +63,15 @@ public ResourceHandlerTests()
...
             observationResource.UpdateCreate = true;
+            observationResource.ConditionalCreate = true;
             observationResource.ConditionalUpdate = true;
             observationResource.Versioning = CapabilityStatement.ResourceVersionPolicy.Versioned;
 
             CapabilityStatementMock.SetupMockResource(_conformanceStatement, ResourceType.Patient, null);
             var patientResource = _conformanceStatement.Rest.First().Resource.Find(x => x.Type == ResourceType.Patient);
             patientResource.ReadHistory = true;
             patientResource.UpdateCreate = true;
+            observationResource.ConditionalCreate = true;
```
Comment (namadabu) [inline src/Microsoft.Health.Fhir.Shared.Core.UnitTests/Features/Resources/ResourceHandlerTests.cs]: `patientResource.ConditionalCreate = true` ?

File: src/Microsoft.Health.Fhir.Shared.Core/Features/Resources/Create/ConditionalCreateResourceHandler.cs [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+        public async Task<UpsertResourceResponse> Handle(ConditionalCreateResourceRequest message, CancellationToken cancellationToken = default)
+        {
+            EnsureArg.IsNotNull(message, nameof(message));
+
+            SearchResult results = await _searchService.SearchAsync(message.Resource.InstanceType, message.ConditionalParameters, cancellationToken);
+
+            int count = results.Results.Count();
+            if (count == 0)
+            {
+                // No matches, no id provided: The server creates the resource
```
Comment (namadabu) [inline src/Microsoft.Health.Fhir.Shared.Core/Features/Resources/Create/ConditionalCreateResourceHandler.cs]: nit: delete `no id provided` since id does not play a role here.

File: test/Microsoft.Health.Fhir.Shared.Tests.E2E/Rest/ConditionalCreateTests.cs [hunk 4]
```diff
@@ -0,0 +1,99 @@
...
+
+            FhirResponse<Observation> updateResponse = await Client.CreateAsync(
+                observation2,
+                $"identifier={identifier}");
+
+            Assert.Equal(HttpStatusCode.OK, updateResponse.StatusCode);
+
+            Observation updatedResource = updateResponse.Resource;
+
+            Assert.Null(updatedResource);
```
Comment (namadabu) [inline test/Microsoft.Health.Fhir.Shared.Tests.E2E/Rest/ConditionalCreateTests.cs]: nit: `Assert.Null(updateResponse.Resource)`

File: src/Microsoft.Health.Fhir.Shared.Core.UnitTests/Features/Resources/ResourceHandlerTests_ConditionalCreate.cs [hunk 5]
```diff
@@ -0,0 +1,95 @@
...
+
+namespace Microsoft.Health.Fhir.Core.UnitTests.Features.Resources
+{
+    /// <summary>
+    /// Tests scenarios for the Conditional Create logic
+    /// </summary>
+    public partial class ResourceHandlerTests
+    {
+        [Fact]
+        public async Task GivenAResource_WhenCreatingConditionallyWithNoIdAndNoExisting_ThenTheServerShouldReturnTheResourceSuccessfully()
```
Comment (namadabu) [inline src/Microsoft.Health.Fhir.Shared.Core.UnitTests/Features/Resources/ResourceHandlerTests_ConditionalCreate.cs]: nit: WhenCreatingConditionallyWithNoExisting

File: src/Microsoft.Health.Fhir.Shared.Core.UnitTests/Features/Resources/ResourceHandlerTests_ConditionalCreate.cs [hunk 6]
```diff
@@ -0,0 +1,95 @@
...
+
+            UpsertResourceResponse result = await _mediator.Send<UpsertResourceResponse>(message);
+
+            Assert.Equal(SaveOutcomeType.Created, result.Outcome.Outcome);
+
+            await _fhirDataStore.Received().UpsertAsync(Arg.Is<ResourceWrapper>(x => x.ResourceId == result.Outcome.Resource.Id), null, true, true, Arg.Any<CancellationToken>());
+        }
+
+        [Fact]
+        public async Task GivenAResourceWithNoId_WhenCreatingConditionallyWithOneMatch_ThenTheServerShouldReturnTheResourceSuccessfully()
```
Comment (namadabu) [inline src/Microsoft.Health.Fhir.Shared.Core.UnitTests/Features/Resources/ResourceHandlerTests_ConditionalCreate.cs]: GivenAResource instead of GivenAResourceWithNoId ?

File: src/Microsoft.Health.Fhir.Shared.Api/Modules/FeatureFlags/ConditionalCreates/ConditionalCreatePostConfigureOptions.cs [hunk 7]
```diff
@@ -0,0 +1,50 @@
...
+// -------------------------------------------------------------------------------------------------
+
+using System.Linq;
+using EnsureThat;
+using Microsoft.AspNetCore.Mvc;
+using Microsoft.Extensions.Options;
+using Microsoft.Health.Fhir.Api.Configs;
+using Microsoft.Health.Fhir.Core.Features.Conformance;
+
+namespace Microsoft.Health.Fhir.Api.Modules.FeatureFlags.ConditionalUpdates
```
Comment (namadabu) [inline src/Microsoft.Health.Fhir.Shared.Api/Modules/FeatureFlags/ConditionalCreates/ConditionalCreatePostConfigureOptions.cs]: `namespace Microsoft.Health.Fhir.Api.Modules.FeatureFlags.ConditionalCreates`

File: src/Microsoft.Health.Fhir.Shared.Api/Modules/FeatureFlags/ConditionalCreates/ConditionalCreateFeatureModule.cs [hunk 8]
```diff
@@ -0,0 +1,29 @@
...
+// -------------------------------------------------------------------------------------------------
+
+using Microsoft.AspNetCore.Mvc;
+using Microsoft.Extensions.DependencyInjection;
+using Microsoft.Extensions.Options;
+using Microsoft.Health.Extensions.DependencyInjection;
+using Microsoft.Health.Fhir.Api.Configs;
+using Microsoft.Health.Fhir.Core.Features.Resources.Create;
+
+namespace Microsoft.Health.Fhir.Api.Modules.FeatureFlags.ConditionalUpdates
```
Comment (namadabu) [inline src/Microsoft.Health.Fhir.Shared.Api/Modules/FeatureFlags/ConditionalCreates/ConditionalCreateFeatureModule.cs]: `namespace Microsoft.Health.Fhir.Api.Modules.FeatureFlags.ConditionalCreates`

File: test/Microsoft.Health.Fhir.Shared.Tests.E2E/Rest/ConditionalCreateTests.cs [hunk 9]
```diff
@@ -0,0 +1,99 @@
...
+            var observation2 = Samples.GetDefaultObservation().ToPoco<Observation>();
+            observation2.Id = Guid.NewGuid().ToString();
+
+            var exception = await Assert.ThrowsAsync<FhirException>(() => Client.CreateAsync(
+                observation2,
+                $"identifier={identifier}"));
+
+            Assert.Equal(HttpStatusCode.PreconditionFailed, exception.Response.StatusCode);
+        }
+    }
```
Comment (namadabu) [inline test/Microsoft.Health.Fhir.Shared.Tests.E2E/Rest/ConditionalCreateTests.cs]: Would it be useful to add a test where we add the If-None-Exist header with an empty value?
